### PR TITLE
support custom import extension

### LINKED ISSUE #245: Raw CSS @imports no longer work
See https://github.com/sass/libsass/releases/tag/3.5.3

1. ~~This changed the current behaviour of libass-python such that importing raw css files without the `.css` extension no longer works. Was this intentional? Might have been useful to include this in the changelog.~~ see below
2. ~~I'm not seeing the deprecation warnings noted above. Maybe I'm not logging warnings or something?~~ see below
3. I want to continue to be able to import raw `css` files after 3.6 is released, will this require any changes to libsass-python? Eg. to support `sass_option_push_import_extension`

### PR DESCRIPTION
**Pull request by samuelcolvin**
fix #245, ref https://github.com/sass/libsass/commit/e690924f83bf726cf259fc11846356a69dcbe491

### CODE REVIEW

File: sass.py [hunk 1]
```diff
@@ -239,7 +239,7 @@ def compile_dirname(
             input_filename = input_filename.encode(fs_encoding)
             s, v, _ = _sass.compile_filename(
                 input_filename, output_style, source_comments, include_paths,
-                precision, None, custom_functions, importers, None,
+                precision, None, custom_functions, importers, None, [],
```
Comment (asottile) [inline sass.py]: `compile_dirname` should probably grow this parameter as well

File: sass.py [hunk 2]
```diff
@@ -584,6 +596,21 @@ def _get_file_arg(key):
             'not {1!r}'.format(SassFunction, custom_functions)
         )
 
+    _custom_exts = kwargs.pop('custom_import_extensions', []) or []
+    if isinstance(_custom_exts, (text_type, bytes)):
+        _custom_exts = [_custom_exts]
```
Comment (asottile) [inline sass.py]: hmm, not sure how I feel about a parameter which ends with `s` that can take a `str`, let's instead raise a `TypeError` and require an iterable of some sort

Also let's just require that they are `text` or ascii compatible:

```python
custom_import_extensions = [
    ext.encode('UTF-8')
    for ext in kwargs.pop('custom_import_extensions', ())
]
```

File: sasstests.py [hunk 3]
```diff
@@ -1409,3 +1409,63 @@ def test_imports_from_cwd(tmpdir):
     with tmpdir.as_cwd():
         out = sass.compile(filename=main_scss.strpath)
         assert out == ''
+
+
+def test_import_no_css(tmpdir):
+    tmpdir.join('other.css').write('body {colour: green}')
```
Comment (asottile) [inline sasstests.py]: `colour`, I like it heh.

File: sasstests.py [hunk 4]
```diff
@@ -1409,3 +1409,63 @@ def test_imports_from_cwd(tmpdir):
...
+
+
+@pytest.mark.parametrize('exts', [
+    '.css',
+    ('.css',),
+    ['.css'],
+    b'.css',
+    [b'.css'],
+    ['.foobar', '.css'],
+    ['.foobar', '.css', b'anything'],
```
Comment (asottile) [inline sasstests.py]: I think this case is unreasonable 🤷‍♂️

**Comment by samuelcolvin** [comment]
Done, except I'm now concerned this is just a sticking plaster over the problem.

It currently causes the warning

```
Including .css files with @import is non-standard behaviour which will be removed in future versions of LibSass.
Use a custom importer to maintain this behaviour. Check your implementations documentation on how to create a custom importer.
```

So this change will be broken again in libsass 3.6.0.

I guess the long-term solution is for libsass-python to provide a css loader. Perhaps we should change the api now to a `allow_css_imports=True/False` which will be a better fit for a future css loader function rather than changing the public API multiple times?

**Comment by asottile** [comment]
@samuelcolvin now that the dust has settled, is this good to go? if so I'll merge and make a release

**Comment by samuelcolvin** [comment]
Yes, I think good to go.

**Comment by asottile** [comment]
just a heads up @samuelcolvin -- looks like upstream has backtracked on the `.css` import situation so we'll be backing this out as well (the option / api from our side will remain but will be a noop):

https://github.com/sass/libsass/pull/2684#event-1712726407

**Comment by samuelcolvin** [comment]
Great news. Very good news they agreed to revert this. 

Do we need to do anything here?

**Comment by xzyfer** [comment]
To summaries the plan moving forward.

It's become clear that people rely on being able to resolve `.css` via `@import`. This is especially the case in Node ecosystem due to packages stylesheets in 3rd party modules. To that point a spec is drafted for officially adding that (behaviour to the language](https://github.com/sass/language/blob/master/proposal/css-imports.md)

Specifically the behaviour we want to deprecate is the using Sass language features in `.css` files. In the future we will produce (de-duped) deprecation warning when we detect a Sass language feature (i.e. nesting, variables, functions, mixins, Sass tyle `@import`s) in .css files.

**Comment by asottile** [comment]
@xzyfer thanks for the great communication on this 👍

**Pull request merged**